anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -0,0 +1,176 @@
+      notEqualAt
+    else if row1.[i] = row2.[i] then
+      // Columns equal, try next one
+      compare notEqualAt (i + 1)
+    else if notEqualAt > -1 then
+      // A non-equal column already exists, abort
+      -1
+    else
+      // Mark current index as not equal and keep going to verify no others are
+      compare i (i + 1)
+
+  compare -1 0
+
+type BucketState =
+  {
+    Aggregators: Aggregator.T array
+    LowCountArguments: Collections.Generic.List<Value list>
+    CountArguments: Collections.Generic.List<Value list>
+    ExecutionContext: ExecutionContext
+    OutputRow: Row
+    mutable IsLowCount: bool
+  }
+
+let private mergeBucket groupingLabelsLength lowCountIndex countIndex destinationBucket fromBucket =
+  let lowCountAggregator = destinationBucket.Aggregators.[lowCountIndex]
+  let countAggregator = destinationBucket.Aggregators.[countIndex]
+
+  fromBucket.LowCountArguments |> Seq.iter lowCountAggregator.Transition
+  fromBucket.CountArguments |> Seq.iter countAggregator.Transition
+
+  destinationBucket.OutputRow.[groupingLabelsLength + lowCountIndex] <-
+    lowCountAggregator.Final destinationBucket.ExecutionContext
+
+  destinationBucket.OutputRow.[groupingLabelsLength + countIndex] <-
+    countAggregator.Final destinationBucket.ExecutionContext
+
+let private LOCAL_AGGREGATION = false
+
+let private executeLed executionContext (childPlan, groupingLabels, aggregators) : seq<Row> =
+  let groupingLabels = Array.ofList groupingLabels
+  let aggFns, aggArgs = aggregators |> Array.ofList |> Executor.Utils.unpackAggregators
+  let groupingLabelsLength = groupingLabels.Length
+  let rowSize = groupingLabelsLength + aggFns.Length
+
+  let lowCountIndex = findAggregator DiffixLowCount aggFns
+  let countIndex = findAggregator DiffixCount aggFns
+
+  let mergeBucketInto = mergeBucket groupingLabelsLength lowCountIndex countIndex
+
+  let makeBucket (group: Value array) =
+    let bucketSeed = Executor.Utils.addValuesToSeed executionContext.NoiseLayers.BucketSeed group
+
+    {
+      Aggregators = aggFns |> Array.map (Aggregator.create executionContext LOCAL_AGGREGATION)
+      LowCountArguments = Collections.Generic.List<Value list>()
+      CountArguments = Collections.Generic.List<Value list>()
+      ExecutionContext =
+        { executionContext with
+            NoiseLayers = { executionContext.NoiseLayers with BucketSeed = bucketSeed }
+        }
+      // Aggregator values are computed later
+      OutputRow = Array.init rowSize (fun i -> if i < groupingLabelsLength then group.[i] else Null)
+      IsLowCount = false
+    }
+
+  let state = Collections.Generic.Dictionary<Row, BucketState>(Row.equalityComparer)
+
+  // Group rows into buckets and keep track of low-count/count arguments
+  for row in Executor.execute executionContext childPlan do
+    let group = groupingLabels |> Array.map (Expression.evaluate row)
+    let argEvaluator = Expression.evaluate row
```

Comment:
Inconsistent usage: either use this in the previous line as well, or expand the call everywhere.

Reply:
Duplicated code from reference :confused:
We need a more generic approach like you said.